Library/src/Multiplayer/EventBus.cpp

```diff
@@ -0,0 +1,185 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "CSP/Multiplayer/EventBus.h"
+
+#include "CSP/Systems/SystemBase.h"
+#include "CallHelpers.h"
+#include "Debug/Logging.h"
+#include "Memory/Memory.h"
+#include "Multiplayer/SignalR/SignalRConnection.h"
+#include "NetworkEventManagerImpl.h"
+
+namespace csp::multiplayer
+{
+
+extern ErrorCode ParseError(std::exception_ptr Exception);
+
+EventBus::~EventBus()
+{
+}
+
+EventBus::EventBus(MultiplayerConnection* InMultiplayerConnection)
+{
+	MultiplayerConnectionInst = InMultiplayerConnection;
+	EventMap				  = {};
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System)
+{
+	if (MultiplayerConnectionInst->Connection == nullptr || !MultiplayerConnectionInst->Connected)
+	{
+		std::string ErrorMessage = "Error: Multiplayer connection is not available.";
+		CSP_LOG_ERROR_FORMAT("%s\n", ErrorMessage.c_str());
+		return;
+	}
+
+	if (!System)
+	{
+		std::string ErrorMessage = "Error: Expected non-null system.";
+		CSP_LOG_ERROR_FORMAT("%s\n", ErrorMessage.c_str());
+		return;
+	}
+
+	if (!EventMap.empty() && EventMap.find(EventName) != EventMap.end())
+	{
+		if (!EventMap[EventName].first.empty())
+		{
+			CSP_LOG_ERROR_FORMAT("Error: there is already at least one callback registered for %s.\n", EventName.c_str());
+			return;
+		}
+		else if (EventMap[EventName].second)
+		{
+			CSP_LOG_ERROR_FORMAT("Error: there is already a system registered for %s.\n", EventName.c_str());
+			return;
+		}
+	}
+
+	EventMap[EventName].second = System;
+}
+
+void EventBus::ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback)
+{
+	if (MultiplayerConnectionInst->Connection == nullptr || !MultiplayerConnectionInst->Connected)
+	{
+		std::string ErrorMessage = "Error: Multiplayer connection is not available.";
+		CSP_LOG_ERROR_FORMAT("%s\n", ErrorMessage.c_str());
+		return;
+	}
+
+	if (!Callback)
+	{
+		std::string ErrorMessage = "Error: Expected non-null callback.";
+		CSP_LOG_ERROR_FORMAT("%s\n", ErrorMessage.c_str());
+		return;
+	}
+
+	if (!EventMap.empty() && EventMap.find(EventName) != EventMap.end() && EventMap[EventName].second)
+	{
+		CSP_LOG_ERROR_FORMAT("Error: there is already a system registered for %s.\n", EventName.c_str());
+		return;
+	}
+
+	EventMap[EventName].first.push_back(Callback);
+}
+
+void EventBus::StopAllListenEvent(const csp::common::String& EventName)
+{
+	if (!EventMap.empty() && EventMap.find(EventName) != EventMap.end())
+	{
+		EventMap.erase(EventName);
+	}
+}
+
+void EventBus::StopSystemListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System)
+{
+	if (System && !EventMap.empty() && EventMap.find(EventName) != EventMap.end() && EventMap[EventName].second == System)
+	{
+		// Remove system
+		EventMap[EventName].second = nullptr;
```

Comment:
Should we actually remove the element here?

Reply:
yeah I don't think this is necessary. the two lines below are enough. I will remove it